### PR TITLE
merge dataset and bucket batching sampler

### PR DESCRIPTION
**Pull request by tkdcjf159**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* dataloader config 추가
    * bucket batching 적용[관련 slack 쓰레드](https://upstageai.slack.com/archives/C08ML7BDTM5/p1746192300574889?thread_ts=1746147157.783179&cid=C08ML7BDTM5) : bucket_ratio 를 0이상 설정하면 bucket batching sampler를 사용합니다.
    * merge_datasets config 추가 : load된 datasets를 하나의 datasets로 concatenate합니다.
    
* train loop 수정
   - 전체 epoch을 통해서 iter를 한번만 초기화 시켜서 여러 epoch에 걸쳐서 중복되는 mini batch data가 나오지 않도록 방지합니다.

### CODE REVIEW

File: torchtune/models/docev/_transform.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def __init__(
         patch_size: int,
         max_num_tiles: int,
         min_num_tiles: int = 1,
-        vision_feature_select_strategy: Literal["default", "full"] = "default",
+        vision_feature_select_strategy: Literal["default", "full"] = "full",
```
Comment (hancheolcho) [inline torchtune/models/docev/_transform.py:74]: 디폴트값이 full 로 변경된 부분을 확인했습니다.

File: torchtune/models/docev/_model_builders.py [hunk 2]
```diff
@@ -110,7 +111,7 @@ def docev_preview(
         connector_type="ldp_v2",
         vision_feature_select_strategy="full",
         tile_size=560,
-        max_num_tiles=9,
+        max_num_tiles=max_num_tiles,
```
Comment (hancheolcho) [inline torchtune/models/docev/_model_builders.py:114]: max_num_tiles를 외부에서 지정가능하도록 변경된 부분을 확인했습니다.

File: recipes/dev/full_finetune_distributed_ufx_dataset.py [hunk 3]
```diff
@@ -782,6 +856,14 @@ def _setup_data(
             if ratio < 1.0:
                 dataset = dataset.select(range(int(len(dataset) * ratio)))
 
+            if merge_datasets:
```
Comment (hancheolcho) [inline recipes/dev/full_finetune_distributed_ufx_dataset.py:859]: merge_datasets = True 일 경우, 데이터셋들을 모두 모아서 하나로 합친 후에 처리하도록 하는 기능이 추가된 것을 확인했습니다.
아래와 같이 코멘트 추가해주면 좋을 것 같습니다.

```suggestion
            if merge_datasets:
                # Collect all datasets and concatenate them before proceeding with further processing
```

File: recipes/dev/full_finetune_distributed_ufx_dataset.py [hunk 4]
```diff
@@ -41,6 +45,76 @@
...
+        ]
+        # 3) 버킷 순서 셔플
+        if self.shuffle:
+            random.seed(self.seed + self.epoch)
+            random.shuffle(buckets)
+        # 4) 각 버킷 내부 셔플
+        for b in buckets:
+            if self.shuffle:
+                random.shuffle(b)
+        # 5) 배치 단위로 쪼개기 → 완전한 배치만
```
Comment (hancheolcho) [inline recipes/dev/full_finetune_distributed_ufx_dataset.py:92]: batch_size == 1 이니 이 부분에서 버려지는 샘플은 없다고 보면 될까요?

File: recipes/dev/full_finetune_distributed_ufx_dataset.py [hunk 5]
```diff
@@ -41,6 +45,76 @@
 log = utils.get_logger("DEBUG")
 
 
+class DistributedBucketSampler(Sampler):
+    """
```
Comment (hancheolcho) [inline recipes/dev/full_finetune_distributed_ufx_dataset.py:49]: 길이가 비슷한 샘플들을 Bucket별로 나눠서 사용한다기보다, 샘플 길이의 오름차순으로 정렬해서 사용하는 방식이군요.
이 때, 정렬 단위가 샘플 1개 단위가 아니라 청크/버켓 단위인 것이구요.

각 gpu가 처리하는 샘플의 시퀀스 길이가 비슷해져서 한쪽이 일하는데 한쪽이 쉬는 경우가 줄어들고, 이 덕분에 전체적인 학습 효율이 증가했다고 이해했습니다.

**Review by hancheolcho — APPROVED** [review]
Bucket Sampler 구현 확인했습니다.
고생 많으셨습니다~
LGTM

File: recipes/dev/full_finetune_distributed_ufx_dataset.py [hunk 6]
```diff
@@ -41,6 +45,76 @@
...
+                random.shuffle(b)
+        # 5) 배치 단위로 쪼개기 → 완전한 배치만
+        batches = []
+        for b in buckets:
+            for i in range(0, len(b), self.batch_size):
+                batch = b[i : i + self.batch_size]
+                if len(batch) == self.batch_size:
+                    batches.append(batch)
+        # 6) 노드별 할당
+        my_batches = batches[self.rank :: self.world_size]
```
Comment (hancheolcho) [inline recipes/dev/full_finetune_distributed_ufx_dataset.py:100]: epoch != 0 일 경우, self.epoch를 쓰면 L86에서 사용하면 bucket 순서가 resume 이전과 이후에 달라져서 이전에 사용된 버킷이 다시 사용될 가능성이 있습니다.
학습이 50% 진행되었을 때 비정상 종료되었다고 가정했을 때, 기존 데이터에서 랜덤하게 샘플/버킷을 선택해 resume한다면, 이전에 사용된 학습 샘플의 50% (전체 학습 데이터의 25%)가 중복되어 사용될 가능성이 있습니다.

L86에서는 seed 값 만으로 random generator를 초기화 하고, L102에 들어가기 이전에 데이터 인덱스 리스트를 아래와 같이 fastforward 해주면 이 문제를 해결할 수 있을 것 같습니다.
```python
num_batches_to_skip = self.epoch * cfg.max_steps_per_epoch * cfg.gradient_accumulation_steps
for batch in my_batches:
    pass
```

File: recipes/dev/full_finetune_distributed_ufx_dataset.py (hunk 4 above)
Comment (tkdcjf159) [reply recipes/dev/full_finetune_distributed_ufx_dataset.py:92]: 넵 batch_size = 1 인 경우는 그렇습니다.

File: recipes/dev/full_finetune_distributed_ufx_dataset.py (hunk 6 above)
Comment (tkdcjf159) [reply recipes/dev/full_finetune_distributed_ufx_dataset.py:100]: 오 감사합니다. 이 부분 참고해서 저도 resume test를 해보고 이상 없는지 확인 후 다른 PR에서 적용하도록 해보겠습니다.
